prusti-viper/src/encoder/spec_encoder.rs

```diff
@@ -225,31 +226,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> SpecEncoder<'p, 'v, 'tcx> {
                                 )
                             );
 
+
                             let sf_post_name = self.encoder.encode_spec_func_name(*def_id, SpecFunctionKind::Post);
+
                             // The result is modeled as the final argument to the post() spec function
-                            let result_name = format!("_{}_forall", vars.args.len() + 2);
+                            let result_var = mir::Local::from_usize(vars.args.len() + 2);
+
+                            // The set of quantified variables
+                            let qvars_post: Vec<_> = vars.args
+                                .iter()
+                                .map(|(arg, arg_ty)|
+                                     self.encode_forall_arg(
+                                         *arg, arg_ty,
+                                         &format!("{}_{}", vars.spec_id, vars.post_id)))
+                                .chain(std::iter::once(
+                                    self.encode_forall_arg(
+                                        result_var, tcx.mk_ty(ty::TyKind::Int(rustc_ast::ast::IntTy::I32)),
+                                        &format!("{}_{}", vars.spec_id, vars.post_id))))
+                                .collect();
+
                             let post_conjunct = vir::Expr::forall(
-                                vars.args
-                                    .iter()
-                                    .map(|(arg, arg_ty)| self.encode_forall_arg(*arg, arg_ty, &format!("{}_{}", vars.spec_id, vars.pre_id)))
-                                    .chain(std::iter::once(vir::LocalVar::new(result_name.clone(), vir::Type::Int)))
-                                    .collect(),
+                                qvars_post.clone(),
                                 vec![], // TODO: encode triggers
                                 vir::Expr::implies(
-                                    encoded_pres,
+                                    // The quantified variables in the precondition have been encoded using
+                                    // different IDs (vars.pre_id vs. vars.post_id), so we need to fix them
```

Comment:
I was in the middle of working on this (should have self-assigned!), but your solution is nicer anyway. I was trying to reencode `encoded_pres` under a different `def_id`. Thanks.

Reply:
Ah, sorry about that! I thought I'd try to help you out a bit on this.

Anyway, the second example in my comment yesterday now also verifies successfully!